staging/src/k8s.io/apiserver/pkg/storage/etcd3/lease_manager.go

```diff
@@ -0,0 +1,207 @@
+func newDefaultLeaseManager(client *clientv3.Client) *leaseManager {
+	return newLeaseManager(client, 8, 60, 0.05)
+}
+
+// creates a new lease manager with the number of buffered leases, lease reuse
+// duration in seconds and percentage. The percentage value x means x*100%.
+func newLeaseManager(client *clientv3.Client, size int, leaseReuseDurationSeconds int64, leaseReuseDurationPercent float64) *leaseManager {
+	return &leaseManager{
+		client: client,
+		size:   size,
+		leaseReuseDurationSeconds: leaseReuseDurationSeconds,
+		leaseReuseDurationPercent: leaseReuseDurationPercent,
+		leaseCache:                make(map[int64]clientv3.LeaseID),
+		leaseLastUsedTime:         make(map[int64]int64),
+		leaseNotification:         make(map[int64][]chan clientv3.LeaseID),
+	}
+}
+
+// SetLeaseReuseDurationSeconds is used for testing purpose. It is used to
+// reduce the extra lease duration to avoid unnecessary timeout in testing.
+func (l *leaseManager) SetLeaseReuseDurationSeconds(duration int64) {
+	l.leaseMu.Lock()
+	l.leaseReuseDurationSeconds = duration
+	l.leaseMu.Unlock()
+}
+
+// GetLease returns a lease based on requested ttl: if a cached lease can be
+// reused, reuse it; otherwise request a new one from etcd.
+func (l *leaseManager) GetLease(ctx context.Context, ttl int64) (clientv3.LeaseID, error) {
+	now := time.Now().Unix()
+	id := l.getLeaseFromCache(now, ttl)
+	if id > clientv3.LeaseID(0) {
+		return id, nil
+	}
+	id, err := l.grantLeaseFromEtcd(ctx, ttl)
+	l.storeLease(id, now, ttl)
+	return id, err
+}
+
+// requests a lease for certain ttl from etcd
+func (l *leaseManager) grantLeaseFromEtcd(ctx context.Context, ttl int64) (clientv3.LeaseID, error) {
+	reuseDurationSeconds := l.getReuseDurationSeconds(ttl)
+	lcr, err := l.client.Lease.Grant(ctx, ttl+reuseDurationSeconds)
+	if err != nil {
+		return clientv3.LeaseID(-1), err
+	}
+	return lcr.ID, nil
+}
+
+// checks whether the requested lease with certain ttl can reuse any of the
+// cached leases: if yes, return the lease id; otherwise, return error
+// indicated by leaseID=-1.
+func (l *leaseManager) getLeaseFromCache(now, ttl int64) clientv3.LeaseID {
+	l.leaseMu.Lock()
+	for ti, id := range l.leaseCache {
+		if ti <= now { // delete expired leases as checking
+			delete(l.leaseCache, ti)
+			delete(l.leaseLastUsedTime, ti)
+		} else {
+			if l.isReusable(ti-now, ttl) { // check if the lease is reusable
+				if id > clientv3.LeaseID(0) { // if reusable and the lease id is available, return it
+					l.leaseLastUsedTime[ti] = now
+					l.leaseMu.Unlock()
+					return id
+				}
+				// if reusable but the lease id is not available,
+				// register a chan for notification and wait for the lease id
+				c := make(chan clientv3.LeaseID, 1)
+				if _, ok := l.leaseNotification[ti]; !ok {
+					l.leaseNotification[ti] = make([]chan clientv3.LeaseID, 0)
+				}
+				l.leaseNotification[ti] = append(l.leaseNotification[ti], c)
+				l.leaseMu.Unlock()
```

Comment:
I'm not a fan of this model of locking - it's very error-prone.

Although, it may look more complicated, the locking would be much simpler, if you would extract most of this method to a separate one, which would return
` func()  clientv3.LeaseID`

And change this code to sth like:
- if you alreayd know the id - return `func() clientv3.LeaseId { return id }`
- if you need to wait for it - return `func() clientv3.LeaseId { id := <-c return id }`

Reply:
fixed